### PR TITLE
[autocast] using new autocast api with device name.

### PR DESCRIPTION
**Pull request by Shan19900305**
Using new autocast APIs with device type name.

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -375,7 +375,7 @@ def forward(
             why_not_sparsity_fast_path = "NestedTensor input is not supported"
         elif mask is not None:
             why_not_sparsity_fast_path = "src_key_padding_mask and mask were both supplied"
-        elif torch.is_autocast_enabled():
+        elif torch.is_autocast_enabled(src.device.type):
```
Comment (albanD) [inline torch/nn/modules/transformer.py:378]: FYI @drisspg the subtle difference here is that the old code was checking only for autocast being enabled on cuda while the updated code will check for autograd being enabled on the device of the input.

**Review by albanD — APPROVED** [review]
Sounds good to me
FYI @drisspg @mikaylagawarecki since this is a non-trivial change in mha

File: torch/nn/modules/transformer.py (hunk 1 above)
Comment (drisspg) [reply torch/nn/modules/transformer.py:378]: I think that should be fine, since there aren't any mixed input devices

**Review by drisspg — APPROVED** [review]

**Comment by Shan19900305** [comment]
I could not reproduced those failed testcases locally， so rebase main branch and push again.

**Comment by albanD** [comment]
cc @drisspg how carefully should we look into torchscript failures vs just skip and ignore for mha?

**Comment by drisspg** [comment]
I think we should investigate these

**Pull request closed without merging**